### PR TITLE
Add new packages to [workspace.members] automatically

### LINKED ISSUE #6378: Feature Request: cargo new --in-workspace [directory] [name]
# The problem

Imagine you're working in a project called `foo` and it's a workspace. You want to create a new member, `bar`, as a subdirectory of `foo`. If you use `cargo new bar` (or `mkdir bar && cd bar && cargo init`) you'll get warnings like this:

```
warning: compiling this new crate may not work due to invalid workspace configuration

current package believes it's in a workspace when it's not:
current:   [new project's Cargo.toml]
workspace: [encompassing workspace's Cargo.toml]

this may be fixable by ensuring that this crate is depended on by the workspace root: 
[encompassing workspace's Cargo.toml]
```

Ideally, that manual work should be automated (if the user wants it).

There's more manual work if you don't want subdirectories. For example, if `foo` is in `code/foo` and you want to make `bar` in `code/bar`, then you have to change the `Cargo.toml` files in *both* directories after using `cargo new`

# The solution

In the case of creating `bar` as a subdirectory, you'd `cd` to the workspace root and use `cargo new --in-workspace . bar` and `foo/Cargo.toml` would be updated to include `bar` in its `[workspace]` section. In the case of creating `bar` outside of the root, you'd use `cargo new --in-workspace ../foo bar` as an example. Similar capabilities should also be given to `cargo init` for consistency

If the directory passed to `--in-workspace` is not a workspace, then there would be an error.

I've never contributed to cargo before (and am pretty new to Rust) but I don't think this is too complicated, and if the idea is accepted I'd love to try and implement it. The changes would take place in [this file](https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/src/cargo/ops/cargo_new.rs). Add a boolean to [NewOptions](https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/src/cargo/ops/cargo_new.rs#L27-L35), change [its constructor](https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/src/cargo/ops/cargo_new.rs#L75-L102), update [this behavior](https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/src/cargo/ops/cargo_new.rs#L607-L614) to change the `workspaces` section of the root's `Cargo.toml` (and the child's `package.workspace` if the child isn't a subdirectory of the root), and it'll be mostly done.

# Notes

This is *not* a breaking change. The new behavior only goes into effect when a new argument (`--in-workspace`) is specified.

Relevant reading: [the original RFC](https://github.com/rust-lang/rfcs/blob/master/text/1525-cargo-workspace.md)

Alternative ideas:
+ `cargo new --root [directory] [name]`
+ `cargo new --member [name]` - strictly for subdirectories (only addresses `foo/bar` example; does not address the issue of creating `code/bar` when your root is `code/foo`)

### PR DESCRIPTION
**Pull request by calavera**
### What does this PR try to resolve?

If a new package is created in a workspace, this change adds the package's path to the workspace's members list automatically.

It doesn't add the package to the list if the path is in the workspace's exclude list, or if the members list doesn't exist already. I noticed that a `cargo_new` test broke if I create the members list when it doesn't exist. This is because the workspace's manifest can be used for package templating. I think it's better to not break that use case.

Fixes #6378

### How should we test and review this PR?

I've included tests in the `cargo_new` suite.

### CODE REVIEW

File: src/cargo/ops/cargo_compile/packages.rs [hunk 1]
```diff
@@ -213,8 +213,3 @@ fn match_patterns(pkg: &Package, patterns: &mut Vec<(glob::Pattern, bool)>) -> b
         is_matched
     })
 }
-
-/// Build [`glob::Pattern`] with informative context.
-pub fn build_glob(pat: &str) -> CargoResult<glob::Pattern> {
```
Comment (epage) [inline src/cargo/ops/cargo_compile/packages.rs:218]: Why was this moved to `restricted_names`?

File: src/cargo/ops/cargo_new.rs [hunk 2]
```diff
@@ -925,3 +941,79 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+    Ok(())
+}
+
+fn matches_in_package_list<'a, P: Iterator<Item = &'a str>>(
+    members: P,
+    package_path: &str,
+) -> CargoResult<bool> {
+    for name in members {
+        let pattern = build_glob(name)?;
+        if pattern.matches(package_path) {
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:996]: `package_path` -> `relative_path` which is the path relative to the CWD which is different than the path relative to the workspace root.
1. Please add a test for this case
2. Please make the test path.  One option is to use `pathdiff` to find the path of the new package relative to the manifest root that was discovered.

File: src/cargo/ops/cargo_new.rs [hunk 3]
```diff
@@ -925,3 +941,79 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
 }
+
+/// Adds the new package member to the [workspace.members] array.
+/// - It first checks if the name matches any element in [workspace.exclude],
+///  and it ignores the name if there is a match.
+/// - Then it check if the name matches any element already in [workspace.members],
+/// and it ignores the name if there is a match.
+/// - If [workspace.members] doesn't exist in the manifest, the new package
+/// is not added to the members array. This prevents manifest files from breaking
+/// if they're only used for package templating.
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:952]: This isn't making sense to me for why we would do this; this is still allowing us to create packages that will fail otherwise.

There might be tests that operate on these failed packages but that was a stop gap until this issue was resolved.

File: src/cargo/ops/cargo_new.rs [hunk 4]
```diff
@@ -925,3 +941,79 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+        .and_then(|workspace| workspace.get("members"))
+        .and_then(|members| members.as_array())
+    {
+        let members = members.iter().map(|s| s.as_str().unwrap());
+        if !matches_in_package_list(members, relative_path)? {
+            let Some(toml_members) = workspace_document["workspace"]["members"].as_array_mut()
+            else {
+                anyhow::bail!("[workspace.members] is not an array");
+            };
+            toml_members.push(relative_path);
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:982]: What formatting situations should we detect and preserve?

File: src/cargo/ops/cargo_new.rs [hunk 5]
```diff
@@ -925,3 +941,79 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+) -> CargoResult<()> {
+    // Don't add the new package to the workspace's members
+    // if there is an exclusion match for it.
+    if let Some(members) = workspace_document
+        .get("workspace")
+        .and_then(|workspace| workspace.get("exclude"))
+        .and_then(|members| members.as_array())
+    {
+        let members = members.iter().map(|s| s.as_str().unwrap());
+        if matches_in_package_list(members, relative_path)? {
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:966]: I don't think we support globs for exclude, see #6009

File: tests/testsuite/cargo_new/add_member_to_workspace/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+[workspace]
```
Comment (epage) [inline tests/testsuite/cargo_new/add_member_to_workspace/Cargo.toml:1]: Why is this file here?

File: tests/testsuite/cargo_new/add_member_to_workspace/empty_members/in/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+[workspace]
```
Comment (epage) [inline tests/testsuite/cargo_new/add_member_to_workspace/empty_members/in/Cargo.toml:1]: > tests/testsuite/cargo_new/add_member_to_workspace/empty_members/in/Cargo.toml

Let's not nest these cases the extra level

File: tests/testsuite/cargo_new/add_member_to_workspace/mod.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+use cargo_test_support::compare::assert_ui;
+use cargo_test_support::curr_dir;
+use cargo_test_support::CargoCommand;
+use cargo_test_support::Project;
+
+#[cargo_test]
```
Comment (epage) [inline tests/testsuite/cargo_new/add_member_to_workspace/mod.rs:6]: Please follow the pattern for these tests with a `mod` in the directory of the test case

File: src/cargo/ops/cargo_new.rs (hunk 2 above)
Comment (epage) [reply src/cargo/ops/cargo_new.rs:996]: See https://github.com/rust-lang/cargo/pull/12779/files#diff-90bbae39fa79149d8bb0faceba6b32e93efbe939964f77c3ad5077a3d0a86942R810

File: src/cargo/ops/cargo_new.rs (hunk 4 above)
Comment (epage) [reply src/cargo/ops/cargo_new.rs:982]: I can't think of any examples but to dig into `toml_edit`s API.

There are likely to main formats to work with
```toml
members = ["one", "two"]
members = [
    "one",   # amount of indentation may vary
    "two",  # use of trailing comma may vary
]
```

With `cargo add`, the only format preserving we do is to keep the list of dependencies sort if it isn't.  We are prescriptive on how to format dependencies rather than making it configurable.

Here, we might have some wiggle room if the people working on `Cargo.toml` style guide (@joshtriplett ?) want to be prescriptive on indentation or trailing commas, because we can just detect specific cases and preserve those and let the user deal with it otherwise.

A third route is a "we don't care" and leave it to a future version of `cargo fmt` that updates `Cargo.toml`.  This is basically the approach that `cargo fix` takes.

**Comment by ehuss** [comment]
r? epage

File: tests/testsuite/cargo_new/add_member_to_workspace/empty_members/in/Cargo.toml (hunk 7 above)
Comment (epage) [reply tests/testsuite/cargo_new/add_member_to_workspace/empty_members/in/Cargo.toml:1]: Please remove the nesting.  So far we have these flat, even for related tests.

File: tests/testsuite/cargo_new/add_member_to_workspace/absolute_package_path/in/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+pub fn add(left: usize, right: usize) -> usize {
```
Comment (epage) [inline tests/testsuite/cargo_new/add_member_to_workspace/absolute_package_path/in/src/lib.rs:1]: Could you pare these tests down
- Remove the README
- Remove the `lib.rs` since the workspace is virtual anyways

(ditto for others)

File: tests/testsuite/cargo_new/not_inherit_workspace_package_table_if_not_members/out/Cargo.toml [hunk 10]
```diff
@@ -1,3 +1,5 @@
+[workspace]
+members = ["foo"]
 [workspace.package]
```
Comment (epage) [inline tests/testsuite/cargo_new/not_inherit_workspace_package_table_if_not_members/out/Cargo.toml:3]: Again, the question we need to answer: how much formatting should we do vs defer to `cargo fmt` like `cargo fix`

In this case, we normally do newlines between tables, but that'd require you to detect when adding `members` affects that.
- In this case, we'd need `members` to have a newline in its decor suffix
- In other cases (like if `resolver = "2"` is present), we'll need to remove the decor suffix on the last field and add it to `members`

File: src/cargo/ops/cargo_new.rs (hunk 4 above)
Comment (epage) [reply src/cargo/ops/cargo_new.rs]: ... that is strange.  I'm the maintainer of `toml_edit` and I'm not sure why that worked.

Comment (epage) [reply src/cargo/ops/cargo_new.rs]: With the idea that we will have `Cargo.toml` auto-fix in the near future, I think it'd be worthwhile doing what auto-fix does for Rust: punt to `cargo fmt`.  Going to check with others to see if we're ok moving forward with that.

Comment (epage) [reply src/cargo/ops/cargo_new.rs]: See https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Automatic.20edits.20to.20.60Cargo.2Ctoml.60

**Comment by epage** [comment]
Could you show the debug representation for each?

**Comment by epage** [comment]
At this point, it comes down to what we want out policy to be for formatting.

**Comment by epage** [comment]
Let's go ahead and remove the auto-formatting but keep the relevant tests.  If nothing else, this is an improvement over what we already have and changing it isn't a breaking change.

File: src/cargo/ops/cargo_new.rs [hunk 11]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+            if Path::new(pat) == relpath {
+                return Ok(());
+            }
+        }
+    }
+
+    let display_path = relpath
+        .to_str()
+        .map(|s| s.replace('\\', "/"))
+        .with_context(|| format!("invalid non-unicode path `{}`", relpath.display()))?;
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:984]: You likely want to do
```rust
let display_path: Result<String, _> = relpath.components().map(|os| os.to_str());
```
The key part being to use `components()` to split the path.  `\\` is a valid character in a path on Linux.

File: src/cargo/ops/cargo_new.rs [hunk 12]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+            let pat = member
+                .as_str()
+                .with_context(|| format!("invalid members' element `{}`", member))?;
+            if Path::new(pat) == relpath {
+                return Ok(());
+            }
+        }
+    }
+
+    let display_path = relpath
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:981]: If we already need to do this, why not move it above and use it for exclude?

File: src/cargo/ops/cargo_new.rs [hunk 13]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+    // in the array already includes the new package's relative path.
+    // - Add the relative path if the members don't match the new package's path.
+    // - Create a new members array if there are no members element in the workspace yet.
+    if let Some(members) = workspace_document
+        .get("workspace")
+        .and_then(|workspace| workspace.get("members"))
+        .and_then(|members| members.as_array())
+    {
+        for member in members {
+            let pat = member.as_str().unwrap();
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:996]: Can we handle this unwrap?

File: src/cargo/ops/cargo_new.rs [hunk 14]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+            let pattern = glob::Pattern::new(pat)
+                .with_context(|| format!("cannot build glob pattern from `{}`", pat))?;
+
+            if pattern.matches_path(&relpath) {
+                return Ok(());
+            }
+        }
+
+        let Some(toml_members) = workspace_document["workspace"]["members"].as_array_mut() else {
+            anyhow::bail!("[workspace.members] is not an array");
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:1006]: Since `workspace.members` is not a table, can you remove the `[]`?  Also, let's use backticks around things like that in messages.

File: src/cargo/ops/cargo_new.rs [hunk 15]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+        };
+
+        toml_members.push(&display_path);
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:1009]: Let's do a sort check on this.  If its sorted, then keep it sorted.  We do that for dependency tables.

File: src/cargo/ops/cargo_new.rs [hunk 16]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+            .as_table_mut()
+            .with_context(|| format!("[workspace] is not a table"))?;
+
+        let mut array = Array::new();
+        array.push(&display_path);
+
+        // only add a trailing new line if the manifest doesn't
+        // include any elements that are not inline tables.
+        // This respects the previous document formatting if there was any already.
+        if !ws.get_values().iter().any(|(_, v)| !v.is_inline_table()) {
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:1021]: Why does something being an inline table matter to this check?  Would we also need to remove a previous trailing newline?

File: src/cargo/ops/cargo_new.rs [hunk 17]
```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
...
+        ws.insert("members", toml_edit::value(array));
+    } else {
+        let mut array = Array::new();
+        array.push(&display_path);
+        array.decor_mut().set_suffix("\n"); // always add a trailing new line
+
+        workspace_document["workspace"]["members"] = toml_edit::value(array);
+    }
+
+    paths::write(&root_manifest_path, workspace_document.to_string())
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:1034]: This should probably be `write_atomic`, see #12744

**Comment by epage** [comment]
While at it, could you clean up the commits for how you'd want them merged into the repo?

**Comment by epage** [comment]
> @epage I don't really have any preference about commit ordering. Do you want me to squash all of them together? What's the project's practice?

If you could clean them up to how you want them in history, that'd be appreciated.  Ideally, all commits would be passing tests, be atomic, and tell a story.  Sometimes, we add a test-only commit and then the commit with the feature/fix so the change in behavior is more obvious.

File: src/cargo/ops/cargo_new.rs [hunk 18]
```diff
@@ -962,11 +962,17 @@ fn update_manifest_with_new_member(
...
-    let display_path = relpath
-        .iter()
-        .filter_map(|comp| comp.to_str())
-        .collect::<Vec<_>>()
-        .join("/");
+    let mut components = Vec::new();
+    for comp in relpath.iter() {
+        let comp = comp.to_str().with_context(|| {
+            format!(
+                "invalid non-ascii component in path `{}`",
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:969]: "invalid unicode component in path"

File: src/cargo/ops/cargo_new.rs [hunk 19]
```diff
@@ -962,11 +962,17 @@ fn update_manifest_with_new_member(
...
     })?;
 
-    let display_path = relpath
-        .iter()
-        .filter_map(|comp| comp.to_str())
-        .collect::<Vec<_>>()
-        .join("/");
+    let mut components = Vec::new();
+    for comp in relpath.iter() {
+        let comp = comp.to_str().with_context(|| {
```
Comment (epage) [inline src/cargo/ops/cargo_new.rs:967]: nit: you can do this as `let components = relpath.iter().map(...).collect::<Result<_, _>>()?;` but its sometimes I pain to do it that way, so not required.

File: src/cargo/util/toml/mod.rs [hunk 20]
```diff
@@ -3470,3 +3470,19 @@ impl<'de> de::Deserialize<'de> for PathValue {
         Ok(PathValue(String::deserialize(deserializer)?.into()))
     }
 }
+
+// Based on Iterator::is_sorted from nightly std; remove in favor of that when stabilized.
+pub fn is_sorted(mut it: impl Iterator<Item = impl PartialOrd>) -> bool {
```
Comment (epage) [inline src/cargo/util/toml/mod.rs:3475]: nit: I think this would more belong in `toml_mut/mod.rs` since its related to toml modifications

**Comment by epage** [comment]
Everything at this point is minor.  Feel free to clean up the commits.

**Comment by epage** [comment]
Thanks!

@bors r+
